appcues/src/main/java/com/appcues/ui/utils/AppcuesAspectRatioModifier.kt

```diff
@@ -0,0 +1,176 @@
+import androidx.compose.ui.unit.isSatisfiedBy
+import com.appcues.data.model.styling.ComponentContentMode
+import kotlin.math.roundToInt
+
+/**
+ * This is an adaptation of the standard Compose AspectRatioModifier, but with logic updates to handle
+ * the matchHeightConstraintsFirst dynamically, based on (1) the contentMode (fit vs fill) and (2) the aspect
+ * ratio of the view being fit within.
+*/
+
+internal class AppcuesAspectRatioModifier(
+    private val aspectRatio: Float,
+    private val contentMode: ComponentContentMode,
+) : LayoutModifier {
+    init {
+        require(aspectRatio > 0) { "aspectRatio $aspectRatio must be > 0" }
+    }
+
+    override fun MeasureScope.measure(
+        measurable: Measurable,
+        constraints: Constraints
+    ): MeasureResult {
+        val size = constraints.findSize()
+        val wrappedConstraints = if (size != IntSize.Zero) {
+            Constraints.fixed(size.width, size.height)
+        } else {
+            constraints
+        }
+        val placeable = measurable.measure(wrappedConstraints)
+        return layout(placeable.width, placeable.height) {
+            placeable.placeRelative(0, 0)
+        }
+    }
+
+    override fun IntrinsicMeasureScope.minIntrinsicWidth(
+        measurable: IntrinsicMeasurable,
+        height: Int
+    ) = if (height != Constraints.Infinity) {
+        (height * aspectRatio).roundToInt()
+    } else {
+        measurable.minIntrinsicWidth(height)
+    }
+
+    override fun IntrinsicMeasureScope.maxIntrinsicWidth(
+        measurable: IntrinsicMeasurable,
+        height: Int
+    ) = if (height != Constraints.Infinity) {
+        (height * aspectRatio).roundToInt()
+    } else {
+        measurable.maxIntrinsicWidth(height)
+    }
+
+    override fun IntrinsicMeasureScope.minIntrinsicHeight(
+        measurable: IntrinsicMeasurable,
+        width: Int
+    ) = if (width != Constraints.Infinity) {
+        (width / aspectRatio).roundToInt()
+    } else {
+        measurable.minIntrinsicHeight(width)
+    }
+
+    override fun IntrinsicMeasureScope.maxIntrinsicHeight(
+        measurable: IntrinsicMeasurable,
+        width: Int
+    ) = if (width != Constraints.Infinity) {
+        (width / aspectRatio).roundToInt()
+    } else {
+        measurable.maxIntrinsicHeight(width)
+    }
+
+    // Suppressing these, as this method is taken from the Compose source and adapted slightly to add the
+    // logic based on contentMode.  Attempting to rewrite this differently would be a more confusing exercise at this point.
+    @Suppress("ComplexMethod", "ReturnCount")
+    private fun Constraints.findSize(): IntSize {
+
+        val containerAspectRatio = this.maxWidth.toDouble() / this.maxHeight.toDouble()
+        val matchHeightConstraintsFirst =
+            (contentMode == ComponentContentMode.FIT && containerAspectRatio > aspectRatio) ||
+                (contentMode == ComponentContentMode.FILL && containerAspectRatio < aspectRatio)
```

Comment:
this is really the only thing changed in this class - adds the desired conditional logic to look at the contentModel FIT vs FILL, then look at the aspect ratio desired and compare to the aspect ratio of the container -- it can then decide which constraint set to prefer to give the expected fit vs fill behavior.